staging/src/k8s.io/client-go/tools/portforward/portforward_test.go

```diff
@@ -165,10 +165,27 @@ func TestParsePortsAndNew(t *testing.T) {
 		// mock-signal the Ready channel
 		close(readyChan)
 
-		if ports, portErr := pf.GetPorts(); portErr != nil {
+		// mock random ports being selected for zero local-ports
+		for i := range pf.ports {
+			if pf.ports[i].Local == 0 {
+				pf.ports[i].Local = uint16(10001 + i)
+			}
+		}
+
+		ports, portErr := pf.GetPorts()
+		if portErr != nil {
 			t.Fatalf("%d: GetPorts: unable to retrieve ports: %s", i, portErr)
-		} else if !reflect.DeepEqual(test.expectedPorts, ports) {
-			t.Fatalf("%d: ports: expected %#v, got %#v", i, test.expectedPorts, ports)
+		} else {
+			for pi, expectedPort := range test.expectedPorts {
+				if e, a := expectedPort.Local, ports[pi].Local; e == 0 && a == 0 {
```

Comment:
If `e==0`, that's a bug in the test case, right? I don't think we need to guard against bugs in test cases.

Reply:
It could also be due to a bug in `GetPorts()`. I think it's a reasonable test because it establishes the contract that a local port of 0 passed in via `ports` when constructing `PortForwarder` must result in a non-zero `.Local` member of the values returned by `GetPorts()`. I think this would true irrespective of how the internal state transformation is done to mock a forwarded port (e.g. if a more elaborate test established actual forwarded ports instead of directly modifying `pf.ports`, I think this test would still hold as valid).

That said, I'm happy to change this if removing it fits better with existing testing conventions of this project.

Reply:
I understand what you said. I agree we need to verify that a non-zero local port is assigned.

My comment is about how the test is written. I would suggest expanding the `expectedPorts` to `expectedParsedPorts` and `expectedGetPorts`, and comparing them to the actual values. The current test is confusing, because `e` is set to 0, while the error message claims the expected value is !0.

Reply:
Ah, my bad for misunderstanding your original post. I agree this is cleaner; thanks!